### ISSUE TITLE
sitewide 508-defect-2 [COGNITION, SCREENREADER]: Pagination SHOULD use meaningful landmarks and aria techniques

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#18398]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:**  Josh

## User Story or Problem Statement

As a screen reader user: 
- I want to know when I'm currently in a pagination
- I want to be able to navigate to the pagination easily using screen reader shortcuts

## Details

Several issues currently exist with the pagination including:
- It is not discoverable by landmark or navigation shortcuts as it is not wrapped in a `nav`.
- If two exist on the page, screen reader users may confuse them with one another as they are not labelled and described appropriately.
- No semantic equivalent to the visual affordance of a current page exists.
- Screen reader users won't know how many items exist in the pagination as pagination links are not written as an unordered list.

I've documented a remediation in a [similar sitewide ticket here](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17085).

## Acceptance Criteria

- [ ] Paginations are wrapped in a `nav` element.
- [ ] Paginations are labeled as paginations using `aria-label`.
- [ ] Paginations are described by their respective table headings using `aria-describedby`.
- [ ] Current page links are noted using `aria-current`.
- [ ] Pagination is written as an unordered list instead of a series of `div` and `a` elements.

## Environment

* Operating System: Mac, Windows
* Browser: Safari, Firefox
* Screenreading device: VoiceOver, NVDA
* Server destination: staging, production

## WCAG or Vendor Guidance

* [Using the aria current attribute - Leonie Watson](https://tink.uk/using-the-aria-current-attribute/)
* [A11y style guide - pagination](https://a11y-style-guide.com/style-guide/section-navigation.html#kssref-navigation-pagination)

## Screenshots or Trace Logs

<img width="967" alt="Screen Shot 2021-01-12 at 11 45 02 AM" src="https://user-images.githubusercontent.com/14154792/104345583-85377d80-54cc-11eb-8041-9a2383ed5b6e.png">

**Comment by joshkimux** [comment]
Not worth writing a ticket over, but here's another weird issue with the pagination. The chevron next to "next" or "previous" will disappear at certain widths/zoom%.
https://user-images.githubusercontent.com/14154792/104388706-c13c0400-5507-11eb-8e5a-084ec0b3070f.mov

**Comment by joshkimux** [comment]
Couple of more issues exist with Firefox/NVDA, I've listed them below, but will need to convert to actual tickets when I have the time. @1Copenut and @noahgelman have y'all caught these before?

On the first results page, the `<h2>` will read "Showing 1-10...", however, when the user clicks on "Next" or "2" and focus is brought back to the newly updated string "Showing 11-20..."-- the _prior_ page's string is announced "Showing 1-10".

<img width="678" alt="Screen Shot 2021-01-12 at 7 08 42 PM" src="https://user-images.githubusercontent.com/14154792/104389537-b2565100-5509-11eb-99f8-2cba19b21cee.png">

Also, NVDA will not announce "Showing 1-10..." when focus is moved there after searching. It will only announce that results are loading.

<img width="677" alt="Screen Shot 2021-01-12 at 7 12 14 PM" src="https://user-images.githubusercontent.com/14154792/104389736-2264d700-550a-11eb-8f1d-d2519494a00a.png">

**Comment by 1Copenut** [comment]
@joshkimux We have captured a few issues for the pagination component in the Vets Design System repo: https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues?q=is%3Aissue+is%3Aopen+pagination+label%3Aaccessibility

Please feel free to log this FF/NVDA issue in that same repo, it does appear to be a bug in the behavior.

**Comment by joshkimux** [comment]
@1Copenut I've submitted a [new bug ticket for the design system #374](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/374) 👍

**Comment by MarciMcGuireGCIO** [comment]
@1Copenut , @joshkimux -- given the last two comments, is this to be considered a sitewide/CMS issue?

**Comment by zacharymorel** [comment]
Hey, nice catch!  This issue here would need to be taken up with Vets Design System team.

**Comment by MarciMcGuireGCIO** [comment]
Moving all outstanding Find a VA Form 508 issues into a single Epic to clean up backlog.

**Comment by zacharymorel** [comment]
For this user story, I'll be focusing on this comment: https://github.com/department-of-veterans-affairs/va.gov-team/issues/18398#issuecomment-759119105

**Comment by zacharymorel** [comment]
@joshkimux In regards to: #18398 (comment) 
After doing some digging, I have figured out that the Select widget, controlled by the design library, 
`aria-live="assertive"` attribute is dominating over the focus on the `<h2>` when the search first gets rendered. 

So the focus results never gets read on the first render because the `aria-live="assertive"` select has priority.

**Comment by zacharymorel** [comment]
I have solved this issue in our particular case by adding an aria-live="polite" to the region of the `<h2>`.
It now will read after the elements given an **Assertive** are read. 

This works, I believe and correct me if I am wrong, because we bring the focus on the `<h2>`, the region attribute sense the change, the screen reader then reads once everything else has been read.

**Comment by zacharymorel** [comment]
@1Copenut Can you this comment https://github.com/department-of-veterans-affairs/va.gov-team/issues/18398#issuecomment-759119105 a review on https://staging.va.gov/find-forms/ ,

**Comment by zacharymorel** [comment]
@joshkimux if you have time, can you give this comment a read: 
https://github.com/department-of-veterans-affairs/va.gov-team/issues/18398#issuecomment-781387887

**Comment by MarciMcGuireGCIO** [comment]
> @joshkimux if you have time, can you give this comment from above a read:
> [#18398 (comment)](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18398#issuecomment-781387887)

@zacharymorel Josh is on PTO until Mar 8 - https://governmentcio-fed.slack.com/archives/CLN0HPZDM/p1612200987004300.

**Comment by zacharymorel** [comment]
Ah thank you @MarciMcGuireGCIO, I'll ping @1Copenut then in regards to the comment.

**Comment by 1Copenut** [comment]
@zacharymorel I took a look and after a few runs through, was able to hear what you were describing. The two aria-live regions are competing so...could we remove the two live regions and make just one on the div that encompasses both the H2 and the select menu? It'd probably need a couple of extra attributes. I'd recommend starting with the following snippet. It might take a couple of rounds of experimenting to get the attributes right for the maximum number of screen readers, but this should be a good starting point.

```html
<div
  aria-atomic="false"
  aria-live="polite"
  aria-relevant="additions text"
  role="region"
>
  <!-- H2, select menu here -->
</div>
```

<img width="1245" alt="Screen Shot 2021-02-26 at 4 31 37 PM" src="https://user-images.githubusercontent.com/934879/109362475-aec82080-7850-11eb-8343-f677ab9f875e.png">

**Comment by zacharymorel** [comment]
@1Copenut After testing heavily this morning and changing various html attributes, I came to this current status within the app:
https://github.com/department-of-veterans-affairs/vets-website/pull/16156/files

I've tested Voice Over on Chrome and Safari. 
Only down side I have notice is after a user changes the sorting select state, it repeats the value selected twice. But everything else _seems_ to be on point? 

![Screen Shot 2021-03-01 at 11 58 03 AM](https://user-images.githubusercontent.com/26075258/109532327-0059f080-7a87-11eb-8902-88811dc2cdb0.png)

**Comment by joshkimux** [comment]
@zacharymorel Don't worry about testing VoiceOver on Chrome as it's always super buggy. If it works on Safari, you're good to go. It's weird that the `aria-live="polite"` is repeating four times; I'll look into this too!

On that note-- @1Copenut , did we look into if the live region is being dynamically added after page load? From deque:

> If you insert the ARIA live region after page load, you must use JavaScript to delay (typically for at least one second) before sending any text into it to make an announcement.

**Comment by zacharymorel** [comment]
@joshkimux if you want to play around/ look, [here](https://github.com/department-of-veterans-affairs/vets-website/pull/16156/files) is the current Draft PR.

**Comment by zacharymorel** [comment]
@joshkimux I am also trying to fix our other [sister 508 issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18389) if I can with this one.

**Comment by joshkimux** [comment]
Thank you for linking these!

**Comment by zacharymorel** [comment]
@joshkimux Where you able to review this?

**Comment by joshkimux** [comment]
@zacharymorel I have not been able to yet! Been really really busy catching up with other tickets during my absence. I'll do my best to get to this tomorrow 🙇

**Comment by joshkimux** [comment]
@zacharymorel Thank you for refreshing me on the ticket! I'll find some time this week to figure out why Firefox/NVDA isn't announcing the `h2`. My hunch is that it's related to `aria-live` being finnicky since it's being added post page load.

**Comment by zacharymorel** [comment]
@joshkimux Can you test Firefox/NVDA if it is still not announcing before you dive in?

**Comment by joshkimux** [comment]
Firefox/NVDA upon hitting "search" will announce "Unavailable," "Loading Search Results," "Loading progress bar" but not the focused `h2` "Showing 1-10 of 20 results for "medical""

**Comment by zacharymorel** [comment]
@MarciMcGuireGCIO I think we should move this as "blocked"

**Comment by MarciMcGuireGCIO** [comment]
@zacharymorel can you provided context for the blocker?  Thanks!

**Comment by zacharymorel** [comment]
I think I need Josh to look deeper when he has time at this.

**Comment by joshkimux** [comment]
@MarciMcGuireGCIO This is completely on me for delaying! I need to find some time to work with @zacharymorel to figure this one out.

**Comment by MarciMcGuireGCIO** [comment]
No worries, @joshkimux - we know you have a full plate.

**Comment by joshkimux** [comment]
@MarciMcGuireGCIO We've determined that this bug is isolated to NVDA. I'll be opening up a new ticket for us to handle this separately!
